src/internal/consistenthashing/etcd.go

```diff
@@ -0,0 +1,291 @@
+			Locks:  map[string]lockInfo{},
+		},
+		prefix:    prefix,
+		stateLock: sync.Mutex{},
+	}
+	etcdCol := collection.NewEtcdCollection(ring.client, path.Join(ring.prefix, "nodes"),
+		nil, &types.BoolValue{}, nil, nil)
+	go ring.watch(ctx, etcdCol)
+	go ring.createNode(ctx, etcdCol, id[0])
+	return ring, nil
+}
+
+// GetIndex returns the index of the Member associated with key. A key belongs to the first Member whose hash is
+// greater than the key. If no Members are greater, then the Ring wraps around to the first Member.
+func (ring *Ring) GetIndex(key string) int {
+	keyHash := ring.hashFn([]byte(key))
+	index := sort.Search(len(ring.members), func(i int) bool {
+		return ring.members[i].Hash >= keyHash
+	})
+	// wrap around to the first node in the Ring.
+	if index >= len(ring.members) {
+		index = 0
+	}
+	return index
+}
+
+// Get is a convenience wrapper around GetIndex that returns the Member that associates to a key.
+func (ring *Ring) Get(key string) Member {
+	return ring.members[ring.GetIndex(key)]
+}
+
+// Lock attempts to lock a key in etcd using consistent hashing. If key does not belong to the ring's node,
+// the call blocks until context is cancelled.
+func (ring *Ring) Lock(ctx context.Context, key string) (context.Context, error) {
+	ring.stateLock.Lock()
+	defer ring.stateLock.Unlock()
+	ticker := time.NewTicker(time.Second * 5)
+	defer ticker.Stop()
+	// todo: this should be event driven instead of by polling for better scalability.
+	for {
+		if len(ring.members) != 0 {
+			index := ring.GetIndex(key)
+			if ring.node.Id == ring.members[index].Id {
+				break
+			}
+		}
+		ring.stateLock.Unlock()
+		select {
+		case <-ctx.Done():
+			ring.stateLock.Lock()
+			return nil, errors.EnsureStack(ctx.Err())
+		case <-ticker.C:
+		}
+		ring.stateLock.Lock()
+	}
+	if info, alreadyExists := ring.node.Locks[key]; alreadyExists {
+		lockCtx := info.Ctx
+		return lockCtx, nil
+	}
+	nodeLock := dlock.NewDLock(ring.client, key)
+	lockCtx, err := nodeLock.Lock(ctx)
+	if err != nil {
+		return nil, errors.EnsureStack(err)
+	}
+	ring.node.Locks[key] = lockInfo{
+		Lock: nodeLock,
+		Ctx:  lockCtx,
+	}
+	ring.logger.WithFields(ring.lockLogFields(key)).Info("claimed lock")
+	return lockCtx, nil
+}
+
+// Unlock attempts to unlock a key in etcd if and only if a ring's node owns a lock reference to it.
+// Remote Member instances must give up their own locks.
+func (ring *Ring) Unlock(key string) error {
+	info, localLock := ring.node.Locks[key]
+	if !localLock {
+		return nil // lock does not exist in this local process.
+	}
+	if err := info.Lock.Unlock(info.Ctx); err != nil {
+		return errors.EnsureStack(err)
+	}
+	delete(ring.node.Locks, key)
+	ring.logger.WithFields(ring.lockLogFields(key)).Info("released lock")
+	return nil
+}
+
+// createNode creates a lease, inserts itself as a key to etcd, keeps the lease alive in the background.
+func (ring *Ring) createNode(ctx context.Context, col collection.EtcdCollection, id string) {
```

Comment:
Would you move this and the watch function closer to the constructor. It improves the readability a bit imo. Generally, I like to have functions like this defined in order and close to where they are used.